### PR TITLE
fix: Do not filter out chunks for store when `From==Through` and `From==start` 

### PR DESCRIPTION
**Pull request by chaudum**
**What this PR does / why we need it**:

This PR fixes a bug where chunks are incorrectly filtered out when their `From` timestamp is equal to their `Through` timestamp (which happens when there is a single log line) and the `From` timestamp is equal to the `from` time of the of the request.

How to reproduce:

1. Insert a single log line with a timestamp exactly at point of an hour
2. Flush ingester
3. Query log line with a split interval of 1h

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

* In a query range, the `From` is inclusive and the `Through` is exclusive
* In a ChunkRef, both `From` and `Through` are inclusive, because `Through` represents the timestamp of the last log line in the chunk.

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [x] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/storage/stores/composite_store_entry.go [hunk 1]
```diff
@@ -90,7 +90,7 @@ func (c *storeEntry) GetChunks(
 func filterForTimeRange(refs []*logproto.ChunkRef, from, through model.Time) []chunk.Chunk {
 	filtered := make([]chunk.Chunk, 0, len(refs))
 	for _, ref := range refs {
-		if through >= ref.From && from < ref.Through {
+		if (through >= ref.From && from < ref.Through) || (ref.From == from && ref.Through == from) {
```
Comment (slim-bean) [inline pkg/storage/stores/composite_store_entry.go:93]:
```suggestion
		// Only include chunks where the query start time (from) is < the chunk end time (ref.Through)
		// and the query end time (through) is >= the chunk start time (ref.From)
		// A special case also exists where a chunk can contain a single log line which results in ref.From being equal to ref.Through, but only include this chunk if this log line timestamp is in the timerange of the query
		if (through >= ref.From && from < ref.Through) || (ref.From == ref.Through && (ref.From >= from && ref.From < through) {
```

File: pkg/storage/stores/composite_store_entry.go [hunk 2]
```diff
@@ -90,7 +90,10 @@ func (c *storeEntry) GetChunks(
 func filterForTimeRange(refs []*logproto.ChunkRef, from, through model.Time) []chunk.Chunk {
 	filtered := make([]chunk.Chunk, 0, len(refs))
 	for _, ref := range refs {
-		if through >= ref.From && from < ref.Through {
+		// Only include chunks where the query start time (from) is < the chunk end time (ref.Through)
+		// and the query end time (through) is >= the chunk start time (ref.From)
+		// A special case also exists where a chunk can contain a single log line which results in ref.From being equal to ref.Through, and that is equal to the from time.
+		if (through >= ref.From && from < ref.Through) || (ref.From == from && ref.Through == from) {
```
Comment (DylanGuedes) [inline pkg/storage/stores/composite_store_entry.go:96]: do you know if there's a relevant difference between using `through.After(x)` vs `through > x`?

Comment (chaudum) [reply pkg/storage/stores/composite_store_entry.go:96]: I don't think so. `through.After()` is an additional function call, but that might get inlined anyway.

Comment (chaudum) [reply pkg/storage/stores/composite_store_entry.go:96]: Quickly checked the following example:


```go
package main

import (
	"fmt"
)

type Time int64

func (t Time) After(other Time) bool {
	return t > other
}

func main() {
	a := Time(1000)
	fmt.Println(a.After(500))
}

```


```console
$ go build -gcflags -m inline.go 
# command-line-arguments
./inline.go:9:6: can inline Time.After
./inline.go:15:21: inlining call to Time.After
./inline.go:15:13: inlining call to fmt.Println
<autogenerated>:1: inlining call to Time.After
./inline.go:15:13: ... argument does not escape
./inline.go:15:21: ~r0 escapes to heap
```

**Review by slim-bean — APPROVED** [review]
LGTM, nice find!

**Pull request merged**